docs/payments/payment-gateway-api.md

```diff
@@ -153,11 +153,13 @@ $order->payment_complete();
 
 This ensures stock reductions are made, and the status is changed to the correct value.
 
-If payment fails, you should throw an error and return null:
+If payment fails, you should throw an error and return an array with the failure status:
 
 ```php
-wc_add_notice( \_\_('Payment error:', 'woothemes') . $error_message, 'error' );
-return;
+wc_add_notice( __('Payment error:', 'woothemes') . $error_message, 'error' );
+return array(
+    'result'   => 'failure',
```

Comment:
Looks good overall, but this array must also contain a `redirect` property:

https://github.com/woocommerce/woocommerce/blob/fe2dd0be2babe6a11f0cfb16f7e2a655624ffcd7/plugins/woocommerce/src/StoreApi/Legacy.php#L71

I'm not sure if this `redirect` would ever get used for failures. Maybe we should conditionally call the line above for non-failures?

I'll let you decide, but if you stick to just docs, please update the other instance in this file as well.

Reply:
> but this array must also contain a redirect property:

Thanks for catching that, @RadoslavGeorgiev! I've added the redirect key in this commit: [c1f4740](https://github.com/woocommerce/woocommerce/pull/53156/commits/c1f4740e800132f56a8f744e46cf32e1ed03bccc).

Additionally, I noticed that the documentation for the process_payment method in the WC_Payment_Gateway abstract class also specifies that the returned array must include a redirect property:

[abstract-wc-payment-gateway.php#L425](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/abstracts/abstract-wc-payment-gateway.php#L425).


>Maybe we should conditionally call the line above for non-failures?

That's a great suggestion! For now, I'll limit this PR to updating the documentation, but I’ll explore your idea in a separate PR.